connector/failoverconnector/failover.go

```diff
@@ -12,11 +17,216 @@ type consumerProvider[C any] func(...component.ID) (C, error)
 type failoverRouter[C any] struct {
 	consumerProvider consumerProvider[C]
 	cfg              *Config
+	pS               *pipelineSelector
+	consumers        []C
+	cancelRetry      *cancelManager
+}
+
+// Manages cancel function for retry goroutine, ends up cleaner than using channels
+type cancelManager struct {
+	lock        sync.Mutex
+	cancelRetry context.CancelFunc
+}
+
+func (m *cancelManager) updateCancelFunc(newCancelFunc context.CancelFunc) {
+	m.lock.Lock()
+	m.cancelRetry = newCancelFunc
+	m.lock.Unlock()
+}
+
+func (m *cancelManager) invokeCancel() {
+	m.lock.Lock()
+	if m.cancelRetry != nil {
+		m.cancelRetry()
+	}
+	m.lock.Unlock()
 }
 
+var (
+	errNoValidPipeline = errors.New("All provided pipelines return errors")
+	errConsumer        = errors.New("Error registering consumer")
+)
+
 func newFailoverRouter[C any](provider consumerProvider[C], cfg *Config) *failoverRouter[C] {
 	return &failoverRouter[C]{
 		consumerProvider: provider,
 		cfg:              cfg,
+		pS:               newPipelineSelector(cfg),
+		cancelRetry:      &cancelManager{},
+	}
+}
+
+func (f *failoverRouter[C]) getCurrentConsumer() C {
+	return f.consumers[f.pS.currentIndex]
+}
+
+func (f *failoverRouter[C]) registerConsumers() error {
+	consumers := make([]C, 0)
+	for _, pipeline := range f.cfg.PipelinePriority {
+		newConsumer, err := f.consumerProvider(pipeline...)
+		if err == nil {
+			consumers = append(consumers, newConsumer)
+		} else {
+			return errConsumer
+		}
+	}
+	f.consumers = consumers
+	return nil
+}
+
+// handlePipelineError evaluates if the error was due to a failed retry or a new failure and will resolve the error
+// by adjusting the priority level accordingly
+func (f *failoverRouter[C]) handlePipelineError() {
+	if f.pS.handleErrorRetryCheck() {
```

Comment:
I'm a bit confused about what this function is intended to check. We know an error occurred, and we need to check whether this requires us to fail over to the next priority level. Is that right?

If so, what does a result of `true` indicate here? That we are failing over?